### ISSUE TITLE
Black screen experiments

### ISSUE THREAD

**Issue by Blaisorblade** [farkam135/GoIV#492]
Some place I can't find, @sarav suggested adding listeners to the screen reader API—`imageReader.setOnImageAvailableListener`. I did (testing on the Nexus 7), and learned a few things:
- currently, `OnImageAvailableListener.onImageAvailable` is invoked rarely—probably, as @sarav guessed, when ImageReader's buffer is full, no more images are posted till we fetch some; (guess in https://github.com/farkam135/GoIV/pull/390#issuecomment-246147320).
- if I add `reader.acquireNextImage().close();`, messages become much more common—basically we get a new notification each time the the screen is updated (flooding the logs)
- either way I got the tablet to crash altogether twice, once with each variant within ~a minute. I don't use it with GoIV all the time, but that doesn't look like a coincidence.

The code I used for the experiment is at
https://github.com/farkam135/GoIV/tree/wip/experiment-setOnImageAvailableListener, though I was testing this on top of #490. #490 alone did not produce crashes (I evaluated that on quite a few pokemons).

I haven't ensured this was repeatable, and I should before we rely on this, but I'm reporting this as a useful data point.

**Comment by sarav** [comment]
I didn't understand your first bullet.

**Comment by sarav** [comment]
2nd bullet -- that's how it's supposed to work. It should be around once every 16 or 32 ms, right? Also, I'd have expected it up work this way without the close() call. What does it even mean? I for got the type of the object returned.

**Comment by Blaisorblade** [comment]
Edited first bullet. 

> It should be around once every 16 or 32 ms, right?

Hm, should have looked at the logs more closely. But I could see moments where the screen was not changed and there was no notification, which is smart.

> Also, I'd have expected it up work this way without the close() call. What does it even mean? I for got the type of the object returned.

Our ImageReader preallocates buffers to store *two* images (as we ask), and each image we get holds onto one of those buffers till we `close` it. At least that's how I get it—part of it is in the docs, part is guesses at how it's implemented. You might want to look at what you wrote in #390.

**Comment by sarav** [comment]
I'm doing all of this on mobile. So, haven't looked at the debug code you put in. So, still a bit confused on what you mean in some of the descriptions. I'll look at it this weekend maybe.

**Comment by Blaisorblade** [comment]
> I'm doing all of this on mobile.

Oh I see, sorry—haven't tried to make my report self-contained. Do take your time.

**Comment by nahojjjen** [comment]
My intuition is telling me that it's the screen-recording that's causing the black screen issue. I can basically not reproduce this issue at all on my phone, so I cant test any changes I do. 

Could someone test either

1. delaying the screen recording if start pokemon go option is enabled

2. start the screen recording in a new thread with low priority, to maybe allow the phone to open pogo first?

**Comment by Blaisorblade** [comment]
@thearaks had the same intuition and IIRC a similar proposal.

I just remembered, the issue became worse when @sarav reduced scans to "only when screen is touched". Since proper screen recorders don't cause problems, it might be that having images in the queue triggers the problem... so maybe scanning more often at startup might also be worth a try.

I'm sorry I can't try this too soon...

**Comment by NightMadness** [comment]
I was able to consistently have a black screen in screenshot mode 
1. Open up GoIV, in screen shot mode, enable confirmation dialog 
2. Take any screenshot, stay in confirmation dialog
3. Open up Pokemon Go
4. Screen is black
5. Click on cancel button, screen comes back to life.

##### Error Log:
```
11-10 23:48:27.181 2503-11722/? E/libEGL: ------------------------------------------------
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI Context(1712054912) Damaged Report-Shader
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI   Shaders(4):
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 0) DAMAGED Handle(10)
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 1) Handle(12)
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 2) Handle(14)
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 3) Handle(16)
11-10 23:48:27.181 2503-11722/? E/libEGL: ------------------------------------------------
11-10 23:48:27.181 2503-11722/? E/libGLESv2: HWUI Protection: wrong calling from app context F:ES2-glDeleteShader, handle(10)
11-10 23:48:27.181 2503-11722/? E/libEGL: ------------------------------------------------
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI Context(1712054912) Damaged Report-Shader
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI   Shaders(4):
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 0) Handle(10)
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 1) DAMAGED Handle(12)
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 2) Handle(14)
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 3) Handle(16)
11-10 23:48:27.181 2503-11722/? E/libEGL: ------------------------------------------------
11-10 23:48:27.181 2503-11722/? E/libGLESv2: HWUI Protection: wrong calling from app context F:ES2-glDeleteShader, handle(12)
11-10 23:48:27.181 2503-11722/? E/libEGL: ------------------------------------------------
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI Context(1712054912) Damaged Report-Shader
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI   Shaders(4):
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 0) Handle(10)
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 1) Handle(12)
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 2) DAMAGED Handle(14)
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 3) Handle(16)
11-10 23:48:27.181 2503-11722/? E/libEGL: ------------------------------------------------
11-10 23:48:27.181 2503-11722/? E/libGLESv2: HWUI Protection: wrong calling from app context F:ES2-glDeleteShader, handle(14)
11-10 23:48:27.181 2503-11722/? E/libEGL: ------------------------------------------------
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI Context(1712054912) Damaged Report-Shader
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI   Shaders(4):
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 0) Handle(10)
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 1) Handle(12)
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 2) Handle(14)
11-10 23:48:27.181 2503-11722/? E/libEGL: HWUI     Unit 3) DAMAGED Handle(16)
11-10 23:48:27.181 2503-11722/? E/libEGL: ------------------------------------------------
11-10 23:48:27.181 2503-11722/? E/libGLESv2: HWUI Protection: wrong calling from app context F:ES2-glDeleteShader, handle(16)
11-10 23:48:27.221 2503-11722/? D/libEGL: eglTerminate EGLDisplay = 0x7f66cbf228
```

##### Screen shot:
![screenshot_20161110-234848](https://cloud.githubusercontent.com/assets/22281828/20204838/18b90bba-a7a1-11e6-9e14-ebfad3f73666.png)

**Comment by Blaisorblade** [comment]
I can confirm the report on a Nexus 5X with Android 7 (though I don't get the HWUI messages). But that should really be a different bug, since there we just have some floating activities on top. And unless there's something wrong with those floating activities, I'm not sure we can fix this. (Or maybe, of course, there is something wrong and we just have no clue).

Also, there's a separate issue:

> Take any screenshot, stay in confirmation dialog

Just tried, that's new—taking non-PoGo-related screenshots IIRC was *NOT* supposed to open that dialog (and IIRC didn't until a while ago). Is that new?

**Comment by nahojjjen** [comment]
Ive tested trying to reproduce the black screen by following the steps mentioned where you take a screenshot on GoIV, but I dont really get a problem when i do so - the screen goes black, but resumes when I close down the overlay.

I've been somewhat successful in reproducing the black-screen issue by 

1. Start GoIV through the notification, which also launches pogo
2. Pause GoIV through notification
3. Go to home screen
4. If no black  screen yet, go to step 1.

I've tried adding a delay to Pokefly startup to allow PoGo to load, which seems to decrease my black screen occurrences when I follow the steps above.  Please test this branch https://github.com/farkam135/GoIV/tree/blackScreen

**Comment by NightMadness** [comment]
was able to replicate with GoIV in screen capture mode
1. open GoIV, start GoIV, scan pokemon, stay in dialog
2. push new debug apk (software will restart)
3. click on start

similar weird error:
```
11-16 00:47:03.995 2501-13542/? E/libEGL: ------------------------------------------------
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI Context(-1757928320) Damaged Report-Shader
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI   Shaders(4):
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 0) DAMAGED Handle(10)
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 1) Handle(12)
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 2) Handle(14)
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 3) Handle(16)
11-16 00:47:03.995 2501-13542/? E/libEGL: ------------------------------------------------
11-16 00:47:03.995 2501-13542/? E/libGLESv2: HWUI Protection: wrong calling from app context F:ES2-glDeleteShader, handle(10)
11-16 00:47:03.995 2501-13542/? E/libEGL: ------------------------------------------------
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI Context(-1757928320) Damaged Report-Shader
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI   Shaders(4):
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 0) Handle(10)
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 1) DAMAGED Handle(12)
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 2) Handle(14)
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 3) Handle(16)
11-16 00:47:03.995 2501-13542/? E/libEGL: ------------------------------------------------
11-16 00:47:03.995 2501-13542/? E/libGLESv2: HWUI Protection: wrong calling from app context F:ES2-glDeleteShader, handle(12)
11-16 00:47:03.995 2501-13542/? E/libEGL: ------------------------------------------------
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI Context(-1757928320) Damaged Report-Shader
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI   Shaders(4):
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 0) Handle(10)
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 1) Handle(12)
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 2) DAMAGED Handle(14)
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 3) Handle(16)
11-16 00:47:03.995 2501-13542/? E/libEGL: ------------------------------------------------
11-16 00:47:03.995 2501-13542/? E/libGLESv2: HWUI Protection: wrong calling from app context F:ES2-glDeleteShader, handle(14)
11-16 00:47:03.995 2501-13542/? E/libEGL: ------------------------------------------------
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI Context(-1757928320) Damaged Report-Shader
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI   Shaders(4):
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 0) Handle(10)
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 1) Handle(12)
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 2) Handle(14)
11-16 00:47:03.995 2501-13542/? E/libEGL: HWUI     Unit 3) DAMAGED Handle(16)
11-16 00:47:03.995 2501-13542/? E/libEGL: ------------------------------------------------
11-16 00:47:03.995 2501-13542/? E/libGLESv2: HWUI Protection: wrong calling from app context F:ES2-glDeleteShader, handle(16)
```

**Comment by nahojjjen** [comment]
> was able to replicate with GoIV in screen capture mode
> 1. open GoIV, start GoIV, scan pokemon, stay in dialog
> 2. push new debug apk (software will restart)
> 3. click on start

So you're saying that pokemon go goes black when you reinstall the apk?

**Comment by nahojjjen** [comment]
The black screen issue was solved in 3.6.0, but reemerged after pokemon go updated. The reasons and reproduction-steps for getting the black-screen bug has changed, and so I'm closing this issue in favor of newer discussion.
